### ISSUE TITLE
How to run multiple times with different personas?

### ISSUE THREAD

**Issue by pcolmer** [postman-solutions-eng/postman-contract-test-generator#21]
Like most APIs, the ones I want to test are based around a CRUD model - create, read, update and delete. To that end, each API has a specific permission requirement, like `create:user` or `delete:user`.

In order to test the APIs thoroughly, I need to be able to run the tests with different personas:

* No permisions at all - 404 should be returned for all calls
* Read-only - 2xx for read APIs, 404 for everything else
* Full - 2xx for all APIs

I've read https://www.readysetcloud.io/blog/allen.helton/how-to-automate-oauth2-token-renewal-in-postman-864420d381a0/ so I can see how I can generate an authentication token for a given persona, but I'm stuck with two challenges:

1. How do I run the contract test generator multiple times, specifying a different persona each time?
2. How do I pass that persona information to the token generation script? Note that we're not using OAuth; we're using a different authorization token and I can retrieve a token for a given persona by passing the appropriate identity information to an API.
3.

**Comment by pcolmer** [comment]
I have had one idea of how this _might_ be implemented but I'm not sure I have got enough coding skills to make the changes myself.

I'm principally thinking of tackling this via `newman` or Postman CLI. For both tools, it is possible to specify the environment details. As such, it would be possible to provide three environments, one for each level of authorization. As well as providing a different authorization token, it would be possible to specify the verbs that should be expected to return a 2xx response code, the rest would then be _expected_ to return a 401 code.

**Comment by jordanwalsh23** [comment]
Yes, you're correct here. Using different environment key-value pairs (or even full environment files) along with the Postman CLI is how you would trigger different scenarios.

You can inject an Authorization header into the collection run, you just need to first set up a variable for it in the Collection level authorization field in Postman. Once you've done that you can send in what ever data you like using the env-var parameter.

Also, the current implementation does not validate at all for security failures. So any requests that are met with a 401 or 403 response are simply considered errors.

If the authentication is successful, but the user role permissions are not adequate to call that particular endpoint, a 403 should be thrown and this again is just considered an error by the testing tool.

I hope that makes sense.
